### PR TITLE
Add SoudenMVDR module

### PR DESCRIPTION
**Pull request by nateanl**
Add a new design of MVDR module.
The `SoudenMVDR` module supports the method proposed by [Souden et, al.](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.725.673&rep=rep1&type=pdf).
The input arguments are:
- multi-channel spectrum.
- PSD matrix of target speech.
- PSD matrix of noise.
- reference channel in the microphone array.
- diagonal_loading option to enable or disable diagonal loading in matrix inverse computation.
- diag_eps for computing the inverse of the matrix.
- eps for computing the beamforming weight.

The output of the module is the single-channel complex-valued spectrum for the enhanced speech.

### CODE REVIEW

**Comment by nateanl** [comment]
Docs: https://output.circle-artifacts.com/output/job/ea1e9c6d-ddf0-4a92-8070-f03a67187951/artifacts/0/docs/transforms.html#soudenmvdr

File: torchaudio/transforms/_transforms.py [hunk 1]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+                is one-hot.
+            diagonal_loading (bool, optional): whether to apply diagonal loading to psd_n
+                (Default: ``True``)
+            diag_eps (float, optional): The coefficient multiplied to the identity matrix for diagonal loading
+                (Default: ``1e-7``)
+            eps (float, optional): a value added to the denominator in the beamforming weight computation.
+                (Default: ``1e-8``)
+
+        Returns:
+            Tensor: Single-channel complex-valued enhanced spectrum of dimension (..., freq, time).
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2142]:
```suggestion
            Tensor: Single-channel complex-valued enhanced spectrum of shape `(..., freq, time)`.
```

File: torchaudio/transforms/_transforms.py [hunk 2]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+        diagonal_loading: bool = True,
+        diag_eps: float = 1e-7,
+        eps: float = 1e-8,
+    ) -> Tensor:
+        """
+        Args:
+            specgram (Tensor): Multi-channel complex-valued spectrum.
+                Tensor of dimension `(..., channel, freq, time)`
+            psd_s (Tensor): The complex-valued power spectral density (PSD) matrix of target speech.
+                Tensor of dimension `(..., freq, channel, channel)`
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2127]: is there a check somewhere that enforces equality between the last two dimensions?

File: torchaudio/transforms/_transforms.py [hunk 3]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+            psd_n (Tensor): The complex-valued power spectral density (PSD) matrix of noise.
+                Tensor of dimension `(..., freq, channel, channel)`
+            reference_channel (int or Tensor): Indicate the reference channel.
+                If the dtype is ``int``, it represent the reference channel index.
+                If the dtype is ``Tensor``, the dimension is `(..., channel)`, where the ``channel`` dimension
+                is one-hot.
+            diagonal_loading (bool, optional): whether to apply diagonal loading to psd_n
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2134]:
```suggestion
            diagonal_loading (bool, optional): If ``True``, enables applying diagonal loading to ``psd_n``.
```

File: torchaudio/transforms/_transforms.py [hunk 4]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+                (Default: ``True``)
+            diag_eps (float, optional): The coefficient multiplied to the identity matrix for diagonal loading
+                (Default: ``1e-7``)
+            eps (float, optional): a value added to the denominator in the beamforming weight computation.
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2138]: nit: nicer to stick to one standard for capitalization
```suggestion
            eps (float, optional): Value to add to the denominator in the beamforming weight formula.
```

File: torchaudio/transforms/_transforms.py [hunk 5]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+        Args:
+            specgram (Tensor): Multi-channel complex-valued spectrum.
+                Tensor of dimension `(..., channel, freq, time)`
+            psd_s (Tensor): The complex-valued power spectral density (PSD) matrix of target speech.
+                Tensor of dimension `(..., freq, channel, channel)`
+            psd_n (Tensor): The complex-valued power spectral density (PSD) matrix of noise.
+                Tensor of dimension `(..., freq, channel, channel)`
+            reference_channel (int or Tensor): Indicate the reference channel.
+                If the dtype is ``int``, it represent the reference channel index.
+                If the dtype is ``Tensor``, the dimension is `(..., channel)`, where the ``channel`` dimension
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2132]:
```suggestion
                If the dtype is ``Tensor``, its shape is `(..., channel)`, where the ``channel`` dimension
```

File: torchaudio/transforms/_transforms.py [hunk 6]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+        """
+        Args:
+            specgram (Tensor): Multi-channel complex-valued spectrum.
+                Tensor of dimension `(..., channel, freq, time)`
+            psd_s (Tensor): The complex-valued power spectral density (PSD) matrix of target speech.
+                Tensor of dimension `(..., freq, channel, channel)`
+            psd_n (Tensor): The complex-valued power spectral density (PSD) matrix of noise.
+                Tensor of dimension `(..., freq, channel, channel)`
+            reference_channel (int or Tensor): Indicate the reference channel.
+                If the dtype is ``int``, it represent the reference channel index.
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2131]:
```suggestion
                If the dtype is ``int``, it represents the reference channel index.
```

File: torchaudio/transforms/_transforms.py [hunk 7]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+        psd_s: Tensor,
+        psd_n: Tensor,
+        reference_channel: Union[int, Tensor],
+        diagonal_loading: bool = True,
+        diag_eps: float = 1e-7,
+        eps: float = 1e-8,
+    ) -> Tensor:
+        """
+        Args:
+            specgram (Tensor): Multi-channel complex-valued spectrum.
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2124]: `Tensor` doesn't seem to be rendering properly as a link in the docs

File: torchaudio/transforms/_transforms.py [hunk 8]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+                Tensor of dimension `(..., channel, freq, time)`
+            psd_s (Tensor): The complex-valued power spectral density (PSD) matrix of target speech.
+                Tensor of dimension `(..., freq, channel, channel)`
+            psd_n (Tensor): The complex-valued power spectral density (PSD) matrix of noise.
+                Tensor of dimension `(..., freq, channel, channel)`
+            reference_channel (int or Tensor): Indicate the reference channel.
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2130]:
```suggestion
            reference_channel (int or Tensor): Specifies the reference channel.
```

File: torchaudio/transforms/_transforms.py [hunk 9]
```diff
@@ -2089,3 +2089,58 @@ def forward(
...
+
+
+class SoudenMVDR(torch.nn.Module):
+    r"""Minimum Variance Distortionless Response (*MVDR* [:footcite:`capon1969high`]) module
+    based on the method proposed by *Souden et, al.* [:footcite:`souden2009optimal`].
+
+    .. devices:: CPU CUDA
+
+    .. properties:: Autograd TorchScript
+
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2101]: perhaps something more to introduce/contextualize the equation? it doesn't read very smoothly right now, as it goes from the intro sentence to the compatibility badges and then suddenly to an equation. also, without probing further, i would have thought that `w_{mvdr}(f)` is what this module returns

File: torchaudio/transforms/_transforms.py (hunk 2 above)
Comment (nateanl) [reply torchaudio/transforms/_transforms.py:2127]: Good point. Adding such check helps users better understand the module usage.

Comment (nateanl) [reply torchaudio/transforms/_transforms.py:2127]: This check is applicable to `F.rtf_power`, `F.rtf_evd`, `F.mvdr_weights_rtf`, `F.mvdr_weights_souden`. It's better to add it in a separate PR.

File: torchaudio/transforms/_transforms.py (hunk 1 above)
Comment (nateanl) [reply torchaudio/transforms/_transforms.py:2142]: In the [CONTRIBUTING.md](https://github.com/pytorch/audio/blob/main/CONTRIBUTING.md) it suggests using `Tensor of audio of dimension (..., time)` or `a tensor of spectrogram with dimensions (..., channel, freq, time)`. 
Which one shall we follow regarding the shape and dimensions?

Comment (hwangjeff) [reply torchaudio/transforms/_transforms.py:2142]: either "shape" or "dimensions" makes sense to me. i think of "dimension" as a scalar value, either representing the number of values in "dimensions", e.g. a tensor of shape (x, y) has a dimension of 2, or as synonymous with "axis", e.g. "along the third dimension"

File: torchaudio/transforms/_transforms.py (hunk 2 above)
Comment (hwangjeff) [reply torchaudio/transforms/_transforms.py:2127]: sure sounds good

File: torchaudio/functional/functional.py [hunk 10]
```diff
@@ -1831,34 +1831,37 @@ def mvdr_weights_souden(
...
             is one-hot.
-        diagonal_loading (bool, optional): whether to apply diagonal loading to psd_n
+        diagonal_loading (bool, optional): If ``True``, enables applying diagonal loading to ``psd_n``.
             (Default: ``True``)
-        diag_eps (float, optional): The coefficient multiplied to the identity matrix for diagonal loading
-            (Default: ``1e-7``)
-        eps (float, optional): a value added to the denominator in mask normalization. (Default: ``1e-8``)
+        diag_eps (float, optional): The coefficient multiplied to the identity matrix for diagonal loading.
+            It is only effective when ``diagonal_loading`` is set to ``True``. (Default: ``1e-7``)
+        eps (float, optional): Value to add to the denominator in the beamforming weight formula.
```
Comment (carolineechen) [inline torchaudio/functional/functional.py:1857]: is `eps` being used?

Comment (nateanl) [reply torchaudio/functional/functional.py:1857]: yep, in line 1867 `ws = numerator / (_compute_mat_trace(numerator)[..., None, None] + eps)`

File: torchaudio/transforms/_transforms.py [hunk 11]
```diff
@@ -2089,3 +2089,66 @@ def forward(
...
+    Given the multi-channel complex-valued spectrum :math:`\textbf{Y}`, the power spectral density (PSD) matrix
+    of target speech :math:`\bf{\Phi}_{\textbf{SS}}`, the PSD matrix of noise :math:`\bf{\Phi}_{\textbf{NN}}`, and
+    a one-hot vector that represents the reference channel :math:`\bf{u}`, the module computes the single-channel
+    complex-valued spectrum of the enhaned speech :math:`\hat{\textbf{S}}`. The formula is defined as:
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2105]:
```suggestion
    complex-valued spectrum of the enhanced speech :math:`\hat{\textbf{S}}`. The formula is defined as:
```

File: torchaudio/transforms/_transforms.py [hunk 12]
```diff
@@ -2089,3 +2089,66 @@ def forward(
...
+        \hat{\textbf{S}}(f) = \textbf{w}_{\text{bf}}(f)^{\mathsf{H}} \textbf{Y}(f)
+
+    where :math:`\textbf{w}_{\text{bf}}(f)` is the MVDR beamforming weight for the :math:`f`-th frequency bin.
+
+    The beamforming weight is computed by:
+
+    .. math::
+        \textbf{w}_{\text{MVDR}}(f) =
+        \frac{{{\bf{\Phi}_{\textbf{NN}}^{-1}}(f){\bf{\Phi}_{\textbf{SS}}}}(f)}
+        {\text{Trace}({{{\bf{\Phi}_{\textbf{NN}}^{-1}}(f) \bf{\Phi}_{\textbf{SS}}}(f))}}\bm{u}
```
Comment (hwangjeff) [inline torchaudio/transforms/_transforms.py:2117]: is "tr" more standard?

```suggestion
        {\text{tr}({{{\bf{\Phi}_{\textbf{NN}}^{-1}}(f) \bf{\Phi}_{\textbf{SS}}}(f))}}\bm{u}
```

**Review by hwangjeff — APPROVED** [review]
lg — just some small things

File: torchaudio/transforms/_transforms.py (hunk 12 above)
Comment (nateanl) [reply torchaudio/transforms/_transforms.py:2117]: I saw different usages in different publications, `trace` in https://www.merl.com/publications/docs/TR2016-072.pdf, `Trace` in https://arxiv.org/pdf/2005.10479.pdf, and `Tr` in https://ieeexplore.ieee.org/abstract/document/7952756.
For a better understanding we can put `Trace` here.
